### PR TITLE
Multi architecture clustering

### PR DESCRIPTION
**Pull request by DBaum1**
For #6380
Registering the native architecture of each server in the nodes database table and updating the generated /1.0 output

### CODE REVIEW

File: lxd/db/node.go [hunk 1]
```diff
@@ -580,6 +582,7 @@ func (c *ClusterTx) NodeOfflineThreshold() (time.Duration, error) {
 // the least number of containers (either already created or being created with
 // an operation).
 func (c *ClusterTx) NodeWithLeastContainers() (string, error) {
+	fmt.Println("NodeWithLeastContainers")
```
Comment (freeekanayaka) [inline lxd/db/node.go]: Leftover print.

File: lxd/api_cluster.go [hunk 2]
```diff
@@ -987,19 +987,17 @@ func tryClusterRebalance(d *Daemon) error {
 	leader, err := d.gateway.LeaderAddress()
 	if err != nil {
 		// This is not a fatal error, so let's just log it.
-		return errors.Wrap(err, "Failed to get current leader member")
+		return errors.Wrap(err, "failed to get current leader member")
```
Comment (freeekanayaka) [inline lxd/api_cluster.go]: The convention in LXD is to use capital case for all error messages, even wrapped one. So this change should be reverted.

File: lxd/api_cluster.go [hunk 3]
```diff
@@ -987,19 +987,17 @@ func tryClusterRebalance(d *Daemon) error {
...
 	}
 	cert := d.endpoints.NetworkCert()
 	client, err := cluster.Connect(leader, cert, true)
 	if err != nil {
-		return errors.Wrap(err, "Failed to connect to leader member")
+		return errors.Wrap(err, "failed to connect to leader member")
```
Comment (freeekanayaka) [inline lxd/api_cluster.go]: Same as above

File: lxd/api_cluster.go [hunk 4]
```diff
@@ -987,19 +987,17 @@ func tryClusterRebalance(d *Daemon) error {
...
 	}
-
 	_, _, err = client.RawQuery("POST", "/internal/cluster/rebalance", nil, "")
 	if err != nil {
-		return errors.Wrap(err, "Request to rebalance cluster failed")
+		return errors.Wrap(err, "request to rebalance cluster failed")
```
Comment (freeekanayaka) [inline lxd/api_cluster.go]: Same as above

File: shared/api/cluster.go [hunk 5]
```diff
@@ -60,4 +60,5 @@ type ClusterMember struct {
 
 	// API extension: clustering_roles
 	Roles []string `json:"roles" yaml:"roles"`
+	Arch    string `json:"arch" yaml:"arch"`
```
Comment (freeekanayaka) [inline shared/api/cluster.go]: We should add and document a new API extension for this field. See ```doc/api-extensions.md``` and ```shared/version/api.go```.

**Comment by freeekanayaka** [comment]
jenkins: test this please

**Comment by freeekanayaka** [comment]
jenkins: test this please

**Comment by freeekanayaka** [comment]
Looks good to me, although it seems jenkins is not going to run tests for a draft PR.

Just to be sure, when you address my comments, please would you amend/rebase your commits, instead of creating new ones on top of them? That keeps the git log history a bit cleaner thanks.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
@freeekanayaka weird, not sure why your jenkins trigger didn't work. We actually wanted to configure it not to test draft PRs but haven't yet found a way :)

Anyway, triggered on the branch so we can get more test results.

**Comment by freeekanayaka** [comment]
@stgraber thanks!

**Comment by freeekanayaka** [comment]
@DBaum1 the test suite fails because of static analysis. You need to run ``gofmt``` against the Go files that you modify.

See for example the output of this run:

https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/7751/arch=amd64,backend=dir,compiler=golang-1.13,restrict=lxd-priv/console

and look for the error messages below the ```==> TEST BEGIN: static analysis``` marker.

**Comment by DBaum1** [comment]
Please ignore the above request, I forgot to fix lxd/db/node.go so it'll still fail

**Comment by stgraber** [comment]
jenkins: test this please

File: lxc/cluster.go [hunk 6]
```diff
@@ -117,7 +117,7 @@ func (c *cmdClusterList) Run(cmd *cobra.Command, args []string) error {
 		if member.Database {
 			database = "YES"
 		}
-		line := []string{member.ServerName, member.URL, database, strings.ToUpper(member.Status), member.Message}
+		line := []string{member.ServerName, member.URL, database, strings.ToUpper(member.Status), member.Message, member.Arch}
```
Comment (stgraber) [inline lxc/cluster.go]: Should be `member.Architecture`, we never abbreviate it in the current API.

File: lxc/cluster.go [hunk 7]
```diff
@@ -128,6 +128,7 @@ func (c *cmdClusterList) Run(cmd *cobra.Command, args []string) error {
 		i18n.G("DATABASE"),
 		i18n.G("STATE"),
 		i18n.G("MESSAGE"),
+		i18n.G("ARCH"),
```
Comment (stgraber) [inline lxc/cluster.go]: I think we'll want `ARCHITECTURE` for that one, I've sent a branch to make `lxc image list` consistent with `lxc list`.

File: lxd/cluster/membership_test.go [hunk 8]
```diff
@@ -203,8 +203,7 @@ func TestAccept_UnmetPreconditions(t *testing.T) {
 			defer gateway.Shutdown()
 
 			c.setup(&membershipFixtures{t: t, state: state})
-
-			_, err := cluster.Accept(state, gateway, c.name, c.address, c.schema, c.api)
+			_, err := cluster.Accept(state, gateway, c.name, c.address, c.schema, c.api, 1)
```
Comment (stgraber) [inline lxd/cluster/membership_test.go]: Please use `osarch.ARCH_64BIT_INTEL_X86` instead of `1`.

File: lxd/cluster/membership_test.go [hunk 9]
```diff
@@ -222,9 +221,8 @@ func TestAccept(t *testing.T) {
 	f := &membershipFixtures{t: t, state: state}
 	f.RaftNode("1.2.3.4:666")
 	f.ClusterNode("1.2.3.4:666")
-
 	nodes, err := cluster.Accept(
-		state, gateway, "buzz", "5.6.7.8:666", cluster.SchemaVersion, len(version.APIExtensions))
+		state, gateway, "buzz", "5.6.7.8:666", cluster.SchemaVersion, len(version.APIExtensions), 1)
```
Comment (stgraber) [inline lxd/cluster/membership_test.go]: Please use `osarch.ARCH_64BIT_INTEL_X86` instead of `1`.

File: lxd/cluster/membership_test.go [hunk 10]
```diff
@@ -306,7 +304,7 @@ func TestJoin(t *testing.T) {
 
 	// Accept the joining node.
 	raftNodes, err := cluster.Accept(
-		targetState, targetGateway, "rusp", address, cluster.SchemaVersion, len(version.APIExtensions))
+		targetState, targetGateway, "rusp", address, cluster.SchemaVersion, len(version.APIExtensions), 1)
```
Comment (stgraber) [inline lxd/cluster/membership_test.go]: Please use `osarch.ARCH_64BIT_INTEL_X86` instead of `1`.

File: lxd/cluster/heartbeat_test.go [hunk 11]
```diff
@@ -142,9 +142,8 @@ func (f *heartbeatFixture) Grow() *cluster.Gateway {
 
 	state, gateway, address := f.node()
 	name := address
-
 	nodes, err := cluster.Accept(
-		targetState, target, name, address, cluster.SchemaVersion, len(version.APIExtensions))
+		targetState, target, name, address, cluster.SchemaVersion, len(version.APIExtensions), 1)
```
Comment (stgraber) [inline lxd/cluster/heartbeat_test.go]: Please use `osarch.ARCH_64BIT_INTEL_X86` instead of `1` and merge the two similar commits together.

File: shared/api/cluster.go [hunk 12]
```diff
@@ -60,4 +60,6 @@ type ClusterMember struct {
 
 	// API extension: clustering_roles
 	Roles []string `json:"roles" yaml:"roles"`
+	// API extension: multi_architecture_clustering
+	Arch string `json:"arch" yaml:"arch"`
```
Comment (stgraber) [inline shared/api/cluster.go]: Need a line break before that new block, also, call it `Architecture` instead of `Arch` and set the json/yaml names to `architecture`.

File: lxd/api_cluster.go [hunk 13]
```diff
@@ -612,7 +613,6 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) response.Response {
 		if err != nil {
 			return err
 		}
-
```
Comment (stgraber) [inline lxd/api_cluster.go]: Please keep the line break.

File: lxd/api_cluster.go [hunk 14]
```diff
@@ -807,14 +807,18 @@ func clusterAcceptMember(
 	client lxd.InstanceServer,
 	name, address string, schema, apiExt int,
 	pools []api.StoragePool, networks []api.Network) (*internalClusterPostAcceptResponse, error) {
-
```
Comment (stgraber) [inline lxd/api_cluster.go]: Please keep the line break.

File: lxd/api_cluster.go [hunk 15]
```diff
@@ -807,14 +807,18 @@ func clusterAcceptMember(
...
+	var arch, err = osarch.ArchitectureGetLocalID()
+	if err != nil {
+		return nil, err
+	}
```
Comment (stgraber) [inline lxd/api_cluster.go]: Please add a line break

File: lxd/api_cluster.go [hunk 16]
```diff
@@ -923,12 +927,10 @@ func clusterNodeDelete(d *Daemon, r *http.Request) response.Response {
 		if err != nil {
 			return response.SmartError(err)
 		}
-
```
Comment (stgraber) [inline lxd/api_cluster.go]: Please keep the line break.

File: lxd/api_cluster.go [hunk 17]
```diff
@@ -923,12 +927,10 @@ func clusterNodeDelete(d *Daemon, r *http.Request) response.Response {
...
 		networks, err := d.cluster.Networks()
 		if err != nil {
 			return response.SmartError(err)
 		}
-
```
Comment (stgraber) [inline lxd/api_cluster.go]: Please keep the line break.

File: lxd/api_cluster.go [hunk 18]
```diff
@@ -941,7 +943,6 @@ func clusterNodeDelete(d *Daemon, r *http.Request) response.Response {
 		if err != nil && err != db.ErrNoSuchObject {
 			return response.SmartError(err)
 		}
-
```
Comment (stgraber) [inline lxd/api_cluster.go]: Please keep the line break.

File: lxd/api_cluster.go [hunk 19]
```diff
@@ -969,7 +970,6 @@ func clusterNodeDelete(d *Daemon, r *http.Request) response.Response {
 		if err != nil {
 			return response.SmartError(err)
 		}
-
```
Comment (stgraber) [inline lxd/api_cluster.go]: Please keep the line break.

File: lxd/api_cluster.go [hunk 20]
```diff
@@ -994,12 +994,10 @@ func tryClusterRebalance(d *Daemon) error {
 	if err != nil {
 		return errors.Wrap(err, "Failed to connect to leader member")
 	}
-
```
Comment (stgraber) [inline lxd/api_cluster.go]: Please keep the line break.

File: lxd/api_cluster.go [hunk 21]
```diff
@@ -994,12 +994,10 @@ func tryClusterRebalance(d *Daemon) error {
...
 	_, _, err = client.RawQuery("POST", "/internal/cluster/rebalance", nil, "")
 	if err != nil {
 		return errors.Wrap(err, "Request to rebalance cluster failed")
 	}
-
```
Comment (stgraber) [inline lxd/api_cluster.go]: Please keep the line break.

File: lxd/api_cluster.go [hunk 22]
```diff
@@ -1071,6 +1069,7 @@ type internalClusterPostAcceptRequest struct {
 	API          int               `json:"api" yaml:"api"`
 	StoragePools []api.StoragePool `json:"storage_pools" yaml:"storage_pools"`
 	Networks     []api.Network     `json:"networks" yaml:"networks"`
+	Arch         int               `json:"arch" yaml:"arch"`
```
Comment (stgraber) [inline lxd/api_cluster.go]: I'd prefer if this also was `Architecture` and `architecture` even if it's less user visible.

File: lxd/api_cluster.go [hunk 23]
```diff
@@ -1162,11 +1161,10 @@ func internalClusterPostRebalance(d *Daemon, r *http.Request) response.Response
 	}
 
 	cert := d.endpoints.NetworkCert()
-	client, err := cluster.Connect(address, cert, true)
+	client, err := cluster.Connect(address, cert, false)
 	if err != nil {
 		return response.SmartError(err)
 	}
-
```
Comment (stgraber) [inline lxd/api_cluster.go]: Please keep the line break.

File: lxd/api_cluster.go [hunk 24]
```diff
@@ -1162,11 +1161,10 @@ func internalClusterPostRebalance(d *Daemon, r *http.Request) response.Response
 	}
 
 	cert := d.endpoints.NetworkCert()
-	client, err := cluster.Connect(address, cert, true)
+	client, err := cluster.Connect(address, cert, false)
```
Comment (stgraber) [inline lxd/api_cluster.go]: Why this change?

File: shared/version/api.go [hunk 25]
```diff
@@ -179,6 +179,7 @@ var APIExtensions = []string{
 	"container_syscall_intercept_mount_fuse",
 	"container_disk_ceph",
 	"virtual-machines",
+	"multi_architecture_clustering",
```
Comment (stgraber) [inline shared/version/api.go]: This should be in the same commit as the one modifying `doc/api-extensions.md`.

**Comment by stgraber** [comment]
Did a quick review, it's mostly code style, naming changes and commit reshuffles at this point.

Looks like the only missing piece here is changing the logic that decides what nodes to put a container on based on the image used and resource availability.

Marking as incomplete for now, please comment once feedback has been addressed.

File: lxd/api_cluster.go (hunk 24 above)
Comment (DBaum1) [reply lxd/api_cluster.go]: Reverted.

**Comment by DBaum1** [comment]
I don't have the simplestreams part of SuitableArchitectures in my branch, so I didn't include those changes for now.

**Comment by DBaum1** [comment]
@stgraber 
Hi, I was wondering if you could go ahead and run the test suite for this PR. The remote simplestreams image server part of SuitableArchitectures hasn't been merged yet which is causing a null pointer error whenever I try `lxc launch ubuntu:18.04. `
Thanks!

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
```
/lxc-ci/build/tmp.CFRUmTbNpO/go/src/github.com/lxc/lxd/lxd/api_cluster.go:810:16: ineffectual assignment to err
```

File: lxd/api_cluster.go [hunk 26]
```diff
@@ -806,13 +807,15 @@ func clusterAcceptMember(
 	name, address string, schema, apiExt int,
 	pools []api.StoragePool, networks []api.Network) (*internalClusterPostAcceptResponse, error) {
 
+	architecture, err := osarch.ArchitectureGetLocalID()
```
Comment (stgraber) [inline lxd/api_cluster.go]: You're never checking that error.

**Comment by freeekanayaka** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
Looking good, can you:
 - Rebase to include those fix commits into the commit they're fixing
 - Rename the extension to `clustering_architecture`
 - Re-title and re-order your commits to something like:
    - `api: Add clustering_architecture extension`
    - `shared/api: Add Architecture to ClusterMember`
    - `lxd/db: Add Architecture to NodeInfo`
    - `lxd/cluster: Track member architecture`
    - `lxc/cluster: Add architecture column in list`

This still leaves you with one more commit to actually make use of all this:
 - `lxd/instances: Use cluster architecture in placement`
